MaxText/layers/pipeline.py

```diff
@@ -162,6 +226,21 @@ def _rotate_right(output_in):
       return jnp.concatenate([last, except_last], axis=0)
     new_shift = _rotate_right(output)
 
+    if self.use_circ_storage:
+      # Insert the circ_storage_mover into new_circ_storage at a microbatch-rotating index.
+      # circ_storage_mover still points to the output of PREVIOUS iteration, which should aid in allowing overlapped compute/async transfers
+      def _rotate_right_and_update(circ_storage_mover_in, circ_storage_in):
+          rotated = _rotate_right(circ_storage_mover_in)
+          rotated = jnp.expand_dims(rotated, 1)
+          # The offset is the previous iterations microbatch ID of the last stage. 
+          offset = (loop_iteration - (self.num_stages - 1) - 1) % self.config.num_pipeline_microbatches # Note extra -1 b/c grabbing from the previous output - using circ_storage_mover before it is updated
+          return jax.lax.dynamic_update_slice_in_dim(circ_storage_in, rotated, offset, axis=1)
```

Comment:
Just checking my understanding: since the last stage is writing dummy data during the first `num_stages` iterations, this offset ensures that stage 0 reads real output when it starts reading from circ_storage.

Reply:
Thats basically right.

We will grab an input from circ_storage  on the num_microbatches index with index given by 
```
circ_storage_batch_idx = loop_iteration % self.config.num_pipeline_microbatches
```

When the first repeat happens (loop_iteration = num_microbatches), this should grab microbatch 0 (and then should grab microbatch 1 in the next loop_iteration, etc). In order to achieve this microbatch 0 should be in index 0 of the num_microbatches dimension of circ_storage. Microbatch 0 will enter the circular_storage when it is finished being processed by the last stage - which happens at iteration #(num_stages - 1), however actually it will enter 1 iteration later than that due to the iteration delay of circ_storage_mover.